File: connectors/connector-msp/src/main/scala/com/taskadapter/connector/msp/TimeCalculator.scala
package com.taskadapter.connector.msp

import com.taskadapter.model.{AllFields, GTask}
import net.sf.mpxj.{Duration, TimeUnit}

object TimeCalculator {
  def calculateTimeAlreadySpent(gTask: GTask): Duration = {
    val doneRatioInPercents = gTask.getValue(AllFields.doneRatio)
    val hours = gTask.getValue(AllFields.estimatedTime)

    val doneRatio = doneRatioInPercents / 100f
    val duration = doneRatio * hours
    Duration.getInstance(duration, TimeUnit.HOURS)
  }

  def calculateRemainingTime(gTask: GTask): Duration = {
    val timeAlreadySpent = calculateTimeAlreadySpent(gTask)
    val estimatedHours = gTask.getValue(AllFields.estimatedTime)
    val hoursLeft = estimatedHours - timeAlreadySpent.getDuration
    Duration.getInstance(hoursLeft, TimeUnit.HOURS)
  }
}

File: connectors/connector-msp/src/main/java/com/taskadapter/connector/msp/TimeCalculator.java
package com.taskadapter.connector.msp;

import com.taskadapter.model.AllFields;
import com.taskadapter.model.GTask;
import net.sf.mpxj.Duration;
import net.sf.mpxj.TimeUnit;

class TimeCalculator {

    static Duration calculateTimeAlreadySpent(GTask gTask) {
        float doneRatioInPercents = getFloatNullSafe(gTask.getValue(AllFields.doneRatio));
        float hours = getFloatNullSafe(gTask.getValue(AllFields.estimatedTime));

        var doneRatio = doneRatioInPercents / 100f;

        var duration = doneRatio * hours;
        return Duration.getInstance(duration, TimeUnit.HOURS);
    }

    static Duration calculateRemainingTime(GTask gTask) {
        var timeAlreadySpent = calculateTimeAlreadySpent(gTask);
        float estimatedHours = getFloatNullSafe(gTask.getValue(AllFields.estimatedTime));
        var hoursLeft = estimatedHours - timeAlreadySpent.getDuration();
        return Duration.getInstance(hoursLeft, TimeUnit.HOURS);
    }

    private static float getFloatNullSafe(Float value) {
        return value == null ? 0 : value;
    }
}
